checkpoint_service.py

```diff
@@ -0,0 +1,71 @@
+import os
+import logging
+
+# List of all objectTypes that we export / import in WM
+USER_OBJECT = "users"
+GROUP_OBJECT = "groups"
+WORKSPACE_NOTEBOOK_OBJECT = "notebooks"
+WORKSPACE_DIRECTORY_OBJECT = "directories"
+WORKSPACE_NOTEBOOK_ACL_OBJECT = "acl_notebooks"
+WORKSPACE_DIRECTORY_ACL_OBJECT = "acl_directories"
+METASTORE_TABLES = "metastore"
+
+# Actions
+WM_EXPORT = "export"
+WM_IMPORT = "import"
+
+class CheckpointKeySet():
+    """Class that deals with checkpoint read and write of a particular object type"""
+
+    """
+    :param checkpoint_enabled: checkpoint enabled / disabled flag
+    :param checkpoint_file: file to read / write object keys for checkpointing
+    """
+    def __init__(self, checkpoint_enabled, checkpoint_file):
+        self._checkpoint_enabled = checkpoint_enabled
+        self._checkpoint_file = checkpoint_file
+        self._checkpoint_key_set = set()
+        self._restore_from_checkpoint_file()
+
+    # writes key to the checkpoint file
+    def write(self, key):
+        if self._checkpoint_enabled and key not in self._checkpoint_key_set:
+            with open(self._checkpoint_file, 'a+') as append_fp:
+                append_fp.write(key + "\n")
```

Comment:
If `checkpoint_key_set` represents the set of keys that are checkpointed (written to the file), we also want to add the key to the _checkpoint_key_set?

Reply:
_checkpoint_key_set is only populated at the start of the export / import of a particular object. Adding keys to the set in the middle doesn't add any value as we are never going to read that key